### PR TITLE
Return an error instead of calling os.Exit()

### PR DESCRIPTION
**Pull request by gvallee**
**Description of the Pull Request (PR):**

Return an error instead of calling os.Exit() in order to facilitate testing. The callers are already checking for errors and appropriately calling os.Exit().

**This fixes or addresses the following GitHub issues:**

- Addresses #3315 

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

**Review by mem — APPROVED** [review]

File: internal/app/singularity/cache_clean_linux.go [hunk 1]
```diff
@@ -65,10 +65,9 @@ func CleanCache(cacheType string) error {
 		err := cache.Clean()
 		return err
 	default:
-		sylog.Fatalf("Not a valid type: %v", cacheType)
-		os.Exit(2)
+		// The caller checks the returned error and will exit as required
+		return fmt.Errorf("Not a valid type: %v", cacheType)
```
Comment (WestleyK) [inline internal/app/singularity/cache_clean_linux.go]: Uncap `N`, and use `%s`.

```suggestion
		return fmt.Errorf("not a valid type: %s", cacheType)
```

File: internal/app/singularity/cache_clean_linux.go [hunk 2]
```diff
@@ -176,8 +175,8 @@ func CleanSingularityCache(cleanAll bool, cacheCleanTypes []string, cacheName st
 		case "all":
 			cleanAll = true
 		default:
-			sylog.Fatalf("Not a valid type: %v", t)
-			os.Exit(2)
+			// The caller checks the returned error and exit when appropriate
+			return fmt.Errorf("Not a valid type: %v", t)
```
Comment (WestleyK) [inline internal/app/singularity/cache_clean_linux.go]: Same here.

File: internal/app/singularity/cache_clean_linux.go [hunk 3]
```diff
@@ -187,8 +186,7 @@ func CleanSingularityCache(cleanAll bool, cacheCleanTypes []string, cacheName st
 			return err
 		}
 		if !foundMatch {
-			sylog.Warningf("No cache found with given name: %v", cacheName)
-			os.Exit(0)
+			return fmt.Errorf("No cache found with given name: %v", cacheName)
```
Comment (WestleyK) [inline internal/app/singularity/cache_clean_linux.go]: Umm, this will exit with a non-0. I think it would be better if you do:

```go
sylog.Warningf("No cache found with given name: %s", cacheName)
return nil
```

Because if a user just wants to make sure that cache is cleared, without failing in there script.

**Review by WestleyK — CHANGES_REQUESTED** [review]
Thanks for fixing this up! I have a couple suggestions...

File: internal/app/singularity/cache_clean_linux.go (hunk 3 above)
Comment (gvallee) [reply internal/app/singularity/cache_clean_linux.go]: fixed

File: internal/app/singularity/cache_clean_linux.go (hunk 2 above)
Comment (gvallee) [reply internal/app/singularity/cache_clean_linux.go]: fixed

File: internal/app/singularity/cache_clean_linux.go (hunk 1 above)
Comment (gvallee) [reply internal/app/singularity/cache_clean_linux.go]: fixed

File: internal/app/singularity/cache_clean_linux.go [hunk 4]
```diff
@@ -187,8 +186,8 @@ func CleanSingularityCache(cleanAll bool, cacheCleanTypes []string, cacheName st
 			return err
 		}
 		if !foundMatch {
-			sylog.Warningf("No cache found with given name: %v", cacheName)
-			os.Exit(0)
+			sylog.Warningf("No cache found with given name: %s", cacheName)
+			return nil
```
Comment (WestleyK) [inline internal/app/singularity/cache_clean_linux.go]: I dont think the `return nil` is necessary... Because it will `return nil` right after the `if !foundMatch {...}`.

**Review by WestleyK — COMMENTED** [review]
One last comment/suggestion. Thanks :+1:

File: internal/app/singularity/cache_clean_linux.go (hunk 4 above)
Comment (gvallee) [reply internal/app/singularity/cache_clean_linux.go]: Good catch. Fixed.

File: internal/app/singularity/cache_clean_linux.go [hunk 5]
```diff
@@ -65,10 +65,9 @@ func CleanCache(cacheType string) error {
 		err := cache.Clean()
 		return err
 	default:
-		sylog.Fatalf("Not a valid type: %v", cacheType)
-		os.Exit(2)
+		// The caller checks the returned error and will exit as required
+		return fmt.Errorf("not a valid type: %s", cacheType)
 	}
-	return nil
```
Comment (WestleyK) [inline internal/app/singularity/cache_clean_linux.go:71]: Why did you remove this line?

Comment (gvallee) [reply internal/app/singularity/cache_clean_linux.go:71]: Yes the linter complained because it is a non-reachable line: all cases of the switch are calling return, including default.

Comment (WestleyK) [reply internal/app/singularity/cache_clean_linux.go:71]: Okay :+1:

**Review by WestleyK — APPROVED** [review]
Looks good to me! Thanks for fixing this up :shipit:
